### PR TITLE
Move from y/N to yes/No for overwriting images

### PR DESCRIPTION
**Pull request by TheBlackKoala**

### CODE REVIEW

File: cotainr/tests/pack/test_conda_install.py [hunk 1]
```diff
@@ -404,15 +404,17 @@ def test_miniforge_still_showing_license(
...
-        assert "conda-forge" in stdout
-        assert "All rights reserved." in stdout
+        miniforgeList = [
+            "Miniforge installer comes",
+            "installing miniforge and is deleted",
+            "bootstrapping executable uses",
+            "micromamba",
+            "conda-forge",
+            "All rights reserved."]
+        for licensePart in miniforgeList:
```
Comment (Chroxvi) [inline cotainr/tests/pack/test_conda_install.py:414]: This doesn't appear to be formatted using [black](https://black.readthedocs.io/en/stable/) as required by the [cotainr style guide](https://cotainr.readthedocs.io/en/stable/development/style_guide.html). Please run black on the entire change set.

**Review by Chroxvi — CHANGES_REQUESTED** [review]
Generally looks good!

Since this is really about making yes/no input consistent, I think we need to align your current changes with the way we ask the user to accept the Miniforge license, i.e.
- [ ] Also rename `val` to `answer` in https://github.com/DeiC-HPC/cotainr/blob/4763532fe6eb5e17c9ca8fe01952cfdd500d5503/cotainr/pack.py#L264 and add a `.lower()` to the it.
- [ ] Add a parametrized test case of "Yes" (in addition to "yes") to https://github.com/DeiC-HPC/cotainr/blob/4763532fe6eb5e17c9ca8fe01952cfdd500d5503/cotainr/tests/pack/test_conda_install.py#L315.

I wonder if we are at the stage where we should refactor this code into a common utility function for handling such yes/no inputs. It is very likely that we are going to ask users for acceptance of other licenses or conditions in the future. What do you think?

File: cotainr/tests/pack/test_conda_install.py (hunk 1 above)
Comment (TheBlackKoala) [reply cotainr/tests/pack/test_conda_install.py:417]: I decided to move the answering to a function that returns a boolean in util instead to make it more standardized.
And blackened everything.

File: cotainr/pack.py [hunk 2]
```diff
@@ -261,8 +262,10 @@ def _display_miniforge_license_for_acceptance(self, *, installer_path):
                 "\n",
             )
             logger.debug(f"The Miniforge displayed license is: {license_text}")
-            val = input(license_text)  # prompt user for acceptance of license terms
-            if val != "yes":
+            answer = util.answer_yes(
+                license_text
+            )  # prompt user for acceptance of license terms
+            if not answer:
```
Comment (akx) [inline cotainr/pack.py:268]: Weird formatting for the comment, and I don't think it really adds much, so...
```suggestion
            if not util.answer_yes(license_text):
```

File: cotainr/util.py [hunk 3]
```diff
@@ -160,3 +166,22 @@ def _print_and_capture_stream(*, stream_handle, print_dispatch):
...
+    Print text and compare the input given to "yes"
+
+    Parameters
+    ----------
+    input_text : str
+        a string to be printed for verification by the user
+
+    Returns
+    -------
+    answer_is_yes : boolean
```
Comment (akx) [inline cotainr/util.py:182]: `boolean` is not a Python type.
```suggestion
    answer_is_yes : bool
```

File: cotainr/util.py [hunk 4]
```diff
@@ -160,3 +166,22 @@ def _print_and_capture_stream(*, stream_handle, print_dispatch):
         captured_stream.append(line)
 
     return captured_stream
+
+
+def answer_yes(input_text):
```
Comment (akx) [inline cotainr/util.py:171]: `answer_yes` seems like a strange name for a function that prompts the user for a yes/no answer?

E.g. [`click` calls it `confirm()`](https://click.palletsprojects.com/en/8.1.x/prompts/#confirmation-prompts).

File: cotainr/util.py [hunk 5]
```diff
@@ -160,3 +166,22 @@ def _print_and_capture_stream(*, stream_handle, print_dispatch):
...
+    input_text : str
+        a string to be printed for verification by the user
+
+    Returns
+    -------
+    answer_is_yes : boolean
+        A boolean indicating whether or not the answer is yes
+    """
+    answer = input(input_text).lower()
+    answer_is_yes = answer == "yes"
```
Comment (akx) [inline cotainr/util.py:186]: This feels pretty user-hostile.

If the user typos `yse`, it'll be the same thing as having answered `no`.

If the only allowed values should be `yes` and `no` (case-insensitive), I feel this should loop until the user enters one of the allowed values.

Comment (Chroxvi) [reply cotainr/util.py:186]: I think that's a fair point - it is a bit harsh to exit and ask the user to start allover if they mistype "yes". And, yes, I know that's what we have done so far - mainly because our initial work on this was somewhat quick-and-dirty. We definitely want the user to clearly answer "yes" before we proceed, but acknowledging a bit of scope creep for this task, I also think we should implement a loop to require either a "yes" or "no" as suggested by @akx.

Comment (TheBlackKoala) [reply cotainr/util.py:186]: @Chroxvi fair. I'll have a look at making more like the usual linux yes/no handling then.

Comment (TheBlackKoala) [reply cotainr/util.py]: The handling has been updated

File: cotainr/pack.py (hunk 2 above)
Comment (TheBlackKoala) [reply cotainr/pack.py]: Though it doesn't seem like a big thing i reckon that the license acceptance is a pivotal point in the program and i'd like to keep the comment as is.

File: cotainr/util.py (hunk 4 above)
Comment (TheBlackKoala) [reply cotainr/util.py]: Name has been changed. I don't think that "confirm" is a very good name

File: cotainr/pack.py [hunk 6]
```diff
@@ -261,8 +262,10 @@ def _display_miniforge_license_for_acceptance(self, *, installer_path):
                 "\n",
             )
             logger.debug(f"The Miniforge displayed license is: {license_text}")
-            val = input(license_text)  # prompt user for acceptance of license terms
-            if val != "yes":
+            answer = util.answer_is_yes(
+                license_text
+            )  # prompt user for acceptance of license terms
+            if not answer:
```
Comment (akx) [inline cotainr/pack.py:268]: I still think
```suggestion
            # prompt user for acceptance of license terms
            if not util.answer_is_yes(license_text):
```
would look better while retaining the comment.

File: cotainr/cli.py [hunk 7]
```diff
@@ -129,11 +129,11 @@ def __init__(
         )
         self.image_path = Path(image_path).resolve()
         if self.image_path.exists():
-            val = input(
+            answer = util.answer_is_yes(
                 f"{self.image_path} already exists. "
-                "Would you like to overwrite it? (y/N) "
-            ).lower()
-            if val != "y":
+                "Would you like to overwrite it? (yes/[N]o) "
```
Comment (akx) [inline cotainr/cli.py:134]: Since `answer_is_yes()` is now responsible for the exact accepted format of the answer, I think the caller shouldn't need to put these in the prompt either.

IOW, `util.answer_is_yes("Is this okay?")` would prompt

```
Is this okay? (yes/[N]o)
```

This also tracks with how that function now changes the prompt after the first query (and includes `Please answer yes/[N]o` internally).

File: cotainr/tests/conftest.py [hunk 8]
```diff
@@ -102,8 +102,13 @@ def factory_mock_input():
...
+        if isinstance(fixed_user_input, list):
+            inputs = iter(fixed_user_input)
+
         def mock_input(prompt):
             print(prompt, end="")
+            if isinstance(fixed_user_input, list):
+                return next(inputs)
             return fixed_user_input
 
         return mock_input
```
Comment (akx) [inline cotainr/tests/conftest.py:114]: It's hard to follow what's happening here, especially since the new behavior (that you can pass in a list of inputs and it will be iterated over (until it runs out)) is not documented.

Also, it's a bit weird to me that the `isinstance(fixed_user_input, list)` case is repeated in the inner and outer functions, and that you read from `inputs`, not `fixed_user_input`...

If it helps, `itertools.repeat()` is a handy way to get an infinitely repeating iterable of a value...

```python
def create_mock_input(value: str | List[str] = ""):
    value_iter = itertools.repeat(value) if isinstance(value, str) else iter(value)

    def mock_input(prompt):
        print(prompt, end="")
        return next(value_iter)

    return mock_input
```

File: cotainr/tests/pack/test_conda_install.py [hunk 9]
```diff
@@ -404,15 +406,16 @@ def test_miniforge_still_showing_license(
...
+        miniforgeList = [
+            "Miniforge installer comes",
+            "installing miniforge and is deleted",
+            "bootstrapping executable uses",
+            "micromamba",
+            "conda-forge",
+            "All rights reserved.",
+        ]
+        for licensePart in miniforgeList:
+            assert (licensePart) in stdout
```
Comment (akx) [inline cotainr/tests/pack/test_conda_install.py:418]: The naming convention for locals around the codebase is `snake_case`, not `camelCase` – and extraneous parens too:
```suggestion
        for license_part in [
            "Miniforge installer comes",
            "installing miniforge and is deleted",
            "bootstrapping executable uses",
            "micromamba",
            "conda-forge",
            "All rights reserved.",
        ]:
            assert license_part in stdout
```

File: cotainr/tests/util/test_answer_is_yes.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+- see the LICENSE file for details.
+
+"""
+
+import pytest
+import cotainr.util
+
+
+class TestAnswerYes:
+    def test_empty(self, capsys, factory_mock_input, monkeypatch):
```
Comment (akx) [inline cotainr/tests/util/test_answer_is_yes.py:15]:
```suggestion
    def test_empty(self, factory_mock_input, monkeypatch):
```

File: cotainr/tests/util/test_answer_is_yes.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        monkeypatch.setattr("builtins.input", factory_mock_input(""))
+        answer = cotainr.util.answer_is_yes("")
+
+        assert not answer
+
+    @pytest.mark.parametrize("outp", ["Test!", "Test\nTest", "Whoop!", "\n", "\nTest!"])
+    @pytest.mark.parametrize("inp", ["n", "nope", "no", "NO!", "Not gonna happen"])
+    def test_n(self, capsys, outp, inp, factory_mock_input, monkeypatch):
```
Comment (akx) [inline cotainr/tests/util/test_answer_is_yes.py:23]: Does this really need to run 5 * 5 = 25 tests to ensure that no matter how the user replies, the prompt printed before they have had the chance to answer is printed correctly? 😄 

(Consider this repeated for the other tests that have these `parametrize` cartesian combinations.)

File: cotainr/tests/util/test_answer_is_yes.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            " ",
+            "O!",
+            "Anbsolutely not",
+            "haha",
+            "I don't know",
+            "yse",
+            "You know it",
+        ],
+    )
+    def test_not_recoqnized(self, capsys, outp, inp, factory_mock_input, monkeypatch):
```
Comment (akx) [inline cotainr/tests/util/test_answer_is_yes.py:47]:
```suggestion
    def test_not_recognized(self, capsys, outp, inp, factory_mock_input, monkeypatch):
```

File: cotainr/tests/util/test_answer_is_yes.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        inputs = [inp, "yes"]
+        monkeypatch.setattr("builtins.input", factory_mock_input(inputs))
+
+        answer = cotainr.util.answer_is_yes(outp)
+        stdout = capsys.readouterr().out
+
+        # Something is wrong in the capturing of stdout when using more than one input
```
Comment (akx) [inline cotainr/tests/util/test_answer_is_yes.py:54]: What does this comment mean?

File: cotainr/util.py [hunk 14]
```diff
@@ -13,6 +13,8 @@
     Run a the command described by `args` while streaming stdout and stderr.
 get_systems()
     Get a dictionary of predefined systems, defined in systems.json
+answer_yes
```
Comment (akx) [inline cotainr/util.py:16]:
```suggestion
answer_is_yes()
```

... but it's debatable whether there's any value in documenting this function separately here in the module docstring 🤷

File: cotainr/cli.py (hunk 7 above)
Comment (Chroxvi) [reply cotainr/cli.py:134]: In addition to this, we need to keep in mind that with the acceptance of the Miniforge license (for which we also use this new function), we just display the Miniforge license text as-is, which includes the final line `Do you accept the license terms? [yes|no]`. One way or the other we need to make this consistent.

File: cotainr/tests/conftest.py (hunk 8 above)
Comment (Chroxvi) [reply cotainr/tests/conftest.py:114]: Maybe we are looking for something along the lines of:
```python
    def create_mock_input(fixed_user_input=None):
        if fixed_user_input is None or isinstance(fixed_user_input, str):
            inputs = iter([fixed_user_input)])
        else:
            inputs = iter(fixed_user_input)

        def mock_input(prompt):
            print(prompt, end="")
            return next(inputs)

        return mock_input
```
with a bit of documentation about what this does

Comment (akx) [reply cotainr/tests/conftest.py:114]: I think `None` is a bad default value (which is why my suggestion above replaces it with an empty string) – after all, a real user with a real keyboard can never get `input()` to return `None`.

@Chroxvi That does a different thing. The previous implementation would always return the same string, yours will return it only once, and then raise `StopIteration`.

File: cotainr/pack.py (hunk 6 above)
Comment (TheBlackKoala) [reply cotainr/pack.py:268]: Done.

File: cotainr/cli.py (hunk 7 above)
Comment (TheBlackKoala) [reply cotainr/cli.py:134]: Done.
answer_is_yes now adds a yes/[N]o prompt to all inputs and i remove the license [yes|no].

File: cotainr/tests/util/test_answer_is_yes.py (hunk 11 above)
Comment (TheBlackKoala) [reply cotainr/tests/util/test_answer_is_yes.py:23]: Done.

File: cotainr/tests/util/test_answer_is_yes.py (hunk 10 above)
Comment (TheBlackKoala) [reply cotainr/tests/util/test_answer_is_yes.py:15]: Done.

File: cotainr/tests/util/test_answer_is_yes.py (hunk 13 above)
Comment (TheBlackKoala) [reply cotainr/tests/util/test_answer_is_yes.py:54]: Not relevant. Removed

File: cotainr/util.py (hunk 14 above)
Comment (TheBlackKoala) [reply cotainr/util.py:16]: Done.

File: cotainr/tests/util/test_answer_is_yes.py (hunk 12 above)
Comment (TheBlackKoala) [reply cotainr/tests/util/test_answer_is_yes.py:47]: Done.

File: cotainr/tests/pack/test_conda_install.py (hunk 9 above)
Comment (TheBlackKoala) [reply cotainr/tests/pack/test_conda_install.py:418]: Done.

File: cotainr/tests/conftest.py (hunk 8 above)
Comment (TheBlackKoala) [reply cotainr/tests/conftest.py:114]: I have given my interpretation of your inputs here. I agree with christians version as the "StopIteration" should not be hit - if it does then there is something wrong in the test any way.

File: cotainr/tests/conftest.py [hunk 15]
```diff
@@ -92,18 +92,25 @@ def data_log_level_names_mapping():
...
+        if isinstance(fixed_user_input, list):
+            inputs = iter(fixed_user_input)
+        else:
+            inputs = iter([fixed_user_input])
+
         def mock_input(prompt):
             print(prompt, end="")
+            if isinstance(fixed_user_input, list):
+                return next(inputs)
             return fixed_user_input
```
Comment (akx) [inline cotainr/tests/conftest.py:114]: I think this implementation is pretty confusing:
* if `fixed_user_input` is a list, assign `inputs` to an iterator over it and use it
* otherwise, assign `inputs` to an iterator over a single instance of the string, *but never use it* since the `if isinstance(fixed_user_input, list):` within `mock_input` is always taken?

File: .gitignore [hunk 16]
```diff
@@ -141,3 +141,6 @@ dmypy.json
 _build/
 
 systems.json
+
+# OSX generated files
+.DS_store
```
Comment (akx) [inline .gitignore:146]: It's worth putting this in your [global git `core.excludesFile`](https://stackoverflow.com/a/22885996/51685) instead.

File: cotainr/cli.py [hunk 17]
```diff
@@ -129,11 +129,10 @@ def __init__(
         )
         self.image_path = Path(image_path).resolve()
         if self.image_path.exists():
-            val = input(
-                f"{self.image_path} already exists. "
-                "Would you like to overwrite it? (y/N) "
-            ).lower()
-            if val != "y":
+            answer = util.answer_is_yes(
+                f"{self.image_path} already exists. " "Would you like to overwrite it?"
```
Comment (akx) [inline cotainr/cli.py:133]:
```suggestion
                f"{self.image_path} already exists. Would you like to overwrite it?"
```

File: cotainr/tests/conftest.py (hunk 15 above)
Comment (TheBlackKoala) [reply cotainr/tests/conftest.py:114]: Ah i overlooked the if-statement. I'll remove it and just return next(inputs)

File: .gitignore (hunk 16 above)
Comment (TheBlackKoala) [reply .gitignore:146]: I don't see any argument against having it there so i'll leave it.

Comment (akx) [reply .gitignore:146]: I would argue this isn't part of the PR :)

The exclude file is a good idea to avoid you having to do this sort of thing for all projects you touch on a Mac.

**Review by Chroxvi — APPROVED** [review]
Looks good to me! I'll merge this now.
